### PR TITLE
[CI] Update dependency installs during test

### PR DESCRIPTION
**Pull request by phschiele**
## Description
The XPRESS installation has failed in master.
Using this opportunity to update package installs according to current availability on PyPI/conda.
Also adds tests for Python 3.12.
Issue link (if applicable): NA

## Type of change
- [ ] New feature (backwards compatible)
- [ ] New feature (breaking API changes)
- [ ] Bug fix
- [x] Other (Documentation, CI, ...)

## [Contribution checklist](https://www.cvxpy.org/contributing/index.html#contribution-checklist)
- [ ] Add our license to new files.
- [ ] Check that your code adheres to our coding style.
- [ ] Write unittests.
- [ ] Run the unittests and check that they’re passing.
- [ ] Run the benchmarks to make sure your change doesn’t introduce a regression.

### CODE REVIEW

File: continuous_integration/install_dependencies.sh [hunk 1]
```diff
@@ -11,41 +11,53 @@ conda config --set remote_backoff_factor 2
...
 fi
 
+if [[ "$RUNNER_OS" != "macos-11" ]]; then
+  if [[ "$PYTHON_VERSION" != "3.12" ]]; then
+    python -m pip install xpress
+  fi
+elif [[ "$PYTHON_VERSION" == "3.11" ]]; then
+  python -m pip install xpress
+fi
+
```
Comment (phschiele) [inline continuous_integration/install_dependencies.sh:60]: Fix for current failure.

File: continuous_integration/install_dependencies.sh [hunk 2]
```diff
@@ -11,41 +11,53 @@ conda config --set remote_backoff_factor 2
...
 elif [[ "$PYTHON_VERSION" == "3.11" ]]; then
     # The earliest version of numpy that works is 1.23.4.
     # Given numpy 1.23.4, the earliest version of scipy we can use is 1.9.3.
-    conda install scipy=1.9.3 numpy=1.23.4 mkl pip pytest openblas ecos scs cvxopt proxsuite daqp "setuptools>65.5.1"
+    conda install scipy=1.9.3 numpy=1.23.4 mkl pip pytest openblas ecos scs osqp clarabel cvxopt proxsuite daqp "setuptools>65.5.1"
+elif [[ "$PYTHON_VERSION" == "3.12" ]]; then
+    # The earliest version of numpy that works is 1.26.0
+    # Given numpy 1.26.0, the earliest version of scipy we can use is 1.9.3.
+    conda install scipy=1.11.3 numpy=1.26.0 mkl pip pytest openblas ecos scs osqp clarabel cvxopt proxsuite daqp "setuptools>65.5.1"
 fi
```
Comment (phschiele) [inline continuous_integration/install_dependencies.sh:31]: Should now work for Python 3.12.

File: continuous_integration/install_dependencies.sh [hunk 3]
```diff
@@ -11,41 +11,53 @@ conda config --set remote_backoff_factor 2
...
-  fi
+if [[ "$PYTHON_VERSION" == "3.12" ]]; then
+  python -m pip install coptpy gurobipy piqp
+elif [[ "$PYTHON_VERSION" == "3.11" ]]; then
+  python -m pip install coptpy gurobipy cplex piqp diffcp "ortools>=9.7,<9.8"
 # Python 3.8 on Windows and Linux will uninstall NumPy 1.16 and install NumPy 1.24 without the exception.
 elif [[ "$PYTHON_VERSION" == "3.8" ]] && [[ "$RUNNER_OS" != "macos-11" ]]; then
-  python -m pip install gurobipy clarabel piqp
+  python -m pip install coptpy gurobipy piqp
 else
```
Comment (PTNobel) [inline continuous_integration/install_dependencies.sh:40]: Why are we removing clarabel from here?

File: continuous_integration/install_dependencies.sh [hunk 4]
```diff
@@ -11,41 +11,53 @@ conda config --set remote_backoff_factor 2
...
-  python -m pip install coptpy cplex diffcp gurobipy xpress clarabel piqp
+  python -m pip install coptpy gurobipy cplex diffcp piqp
```
Comment (PTNobel) [inline continuous_integration/install_dependencies.sh:41]: Ditto re why remove clarabel

**Review by PTNobel — COMMENTED** [review]
Couple questions; I might be missing something

File: continuous_integration/install_dependencies.sh (hunk 3 above)
Comment (phschiele) [reply continuous_integration/install_dependencies.sh:40]: Moved it to the conda install, but that created issues. Moved it back in the latest commit.

File: continuous_integration/install_dependencies.sh (hunk 1 above)
Comment (phschiele) [reply continuous_integration/install_dependencies.sh:60]: @PTNobel Any idea why this is installing `xpress` for:
```
    RUNNER_OS: macos-11
    PYTHON_VERSION: 3.10
```

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -200,7 +200,6 @@ jobs:
         shell: bash
         run: |
           # Mac has a different syntax for sed -i, this works across oses
-          sed -i.bak -e "s/name='cvxpy',/name='cvxpy-base',/g" pyproject.toml
```
Comment (phschiele) [inline .github/workflows/build.yml:203]: Apparently, this line didn't do anything after some changes in `pyproject.toml`. Turns out it is not needed. This line does not show up in the build files.

**Comment by phschiele** [comment]
@PTNobel @SteveDiamond any idea why xpress is still trying to be installed on mac even though I have:
```
if [[ "$PYTHON_VERSION" == "3.11" ]] && [[ "$RUNNER_OS" != "macos-11" ]]; then
  python -m pip install xpress
fi
```
and the env is
```
  env:
    RUNNER_OS: macos-11
```

**Review by SteveDiamond — APPROVED** [review]

**Pull request merged**